### PR TITLE
SNOW-1569005: Deduplicate queries and post actions with building queries from binary plan

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1569005

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   When building plans from binary or set statements, we currently simple merge all queries. This PR fixes by deduplicating the queries.

### CODE REVIEW

**Review by sfc-gh-jdu — APPROVED** [review]

**Review by sfc-gh-jrose — APPROVED** [review]

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 1]
```diff
@@ -637,8 +637,15 @@ def build_binary(
                 if post_action not in post_actions:
                     post_actions.append(copy.copy(post_action))
         else:
-            merged_queries = select_left.queries[:-1] + select_right.queries[:-1]
-            post_actions = select_left.post_actions + select_right.post_actions
+            merged_queries = select_left.queries[:-1]
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:640]: I think we can merge the code for query merge here for the if else branch, we only need a parameter protection for referenced_ctes

File: src/snowflake/snowpark/_internal/compiler/query_generator.py [hunk 2]
```diff
@@ -89,8 +89,12 @@ def generate_queries(
             plan_queries = get_snowflake_plan_queries(
                 snowflake_plan, self.resolved_with_query_block
             )
-            queries.extend(plan_queries[PlanQueryType.QUERIES])
-            post_actions.extend(plan_queries[PlanQueryType.POST_ACTIONS])
+            for query in plan_queries[PlanQueryType.QUERIES]:
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/query_generator.py:92]: I don't think we will need this, the queries and post actions for each plan should be disjoint

**Review by sfc-gh-yzou — COMMENTED** [review]
@sfc-gh-aalam can you trigger the snowpark daily test with your pr to see if it passed with sql simplifier off

File: src/snowflake/snowpark/_internal/compiler/query_generator.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/query_generator.py:92]: without this change, when `_query_compilation_stage_enabled=True` and `cte_optimization_enabled=True`, duplicate queries re-arise.

**Comment by sfc-gh-aalam** [comment]
> @sfc-gh-aalam can you trigger the snowpark daily test with your pr to see if it passed with sql simplifier off

https://github.com/snowflakedb/snowpark-python/actions/runs/10395188222

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:640]: referenced_ctes are only enabled when we have `_query_compilation_stage_enabled=True` right. This is the False branch.
What kind of protection are you thinking of exaclty.

File: src/snowflake/snowpark/_internal/compiler/query_generator.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/query_generator.py:92]: I see your point. You are right, we shouldn't de-duplicate at this point. I think there is another bug which is adding duplicate queries with new cte compilation stage. I'll try to find it.

File: tests/integ/modin/frame/test_apply.py [hunk 3]
```diff
@@ -702,7 +702,7 @@ def foo(row):
     ],
 )
 @sql_count_checker(
-    query_count=18,
+    query_count=13,
```
Comment (sfc-gh-aalam) [inline tests/integ/modin/frame/test_apply.py:705]: without dedup
```
0. ls '@"TESTDB_SNOWPARK_PYTHON"."TESTSCHEMA_SNOWPARK_PYTHON".SNOWPARK_TEMP_STAGE_6VYBSG222X'
1. SELECT "name" FROM ( SELECT  *  FROM  TABLE ( RESULT_SCAN('01b67b2d-0001-9f54-0000-047d100fa0ea')))
2. CREATE TEMPORARY  FUNCTION  "TESTDB_SNOWPARK_PYTHON"."TESTSCHEMA_SNOWPARK_PYTHON".SNOWPARK_TEMP_TABLE_FUNCTION_28BQP90FQN(a...
3. CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_44LOL2QE6K("__index__" BIGINT, "A" DOUBLE, "B" DOUBLE, "C" DOUBLE, "__row_position__" BIGINT)
4. INSERT  INTO SNOWPARK_TEMP_TABLE_44LOL2QE6K("__index__", "A", "B", "C", "__row_position__") VALUES (?, ?, ?, ?, ?)
5. CREATE  TEMPORARY  TABLE  "TESTDB_SNOWPARK_PYTHON"."TESTSCHEMA_SNOWPARK_PYTHON"."SNOWPARK_TEMP_TABLE_LE4JZ1D6QF"("__row_position__" BIGINT, "LABEL" STRING(16777216), "VALUE" VARIANT)   AS  SELECT  *  FROM ( SELECT "__row_position__", "LABEL", "VALUE" FROM ( SELECT T_RIGHT."__row_position__", T_RIGHT."LABEL", T_RIGHT."VALUE" FROM ( SELECT "__index__", "A", "B", "C", "__row_position__", round(("__row_position__" / 1000 :: INT), 0) AS "partition_id" FROM ( SELECT  *  FROM (SNOWPARK_TEMP_TABLE_44LOL2QE6K))) AS T_LEFT JOIN  TABLE ("TESTDB_SNOWPARK_PYTHON"."TESTSCHEMA_SNOWPARK_PYTHON".SNOWPARK_TEMP_TABLE_FUNCTION_28BQP90FQN("__row_position__", "A", "B", "C")  OVER (PARTITION BY "partition_id" )) AS T_RIGHT))
6. DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_44LOL2QE6K
7. CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_44LOL2QE6K("__index__" BIGINT, "A" DOUBLE, "B" DOUBLE, "C" DOUBLE, "__row_position__" BIGINT)
8. alter session set query_tag = '  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/frame.py", line 706, in num_rows\n    return count_rows(self.ordered_dataframe)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/utils.py", line 1668, in count_rows\n    df.select(df.row_count_snowflake_quoted_identifier)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/_internal/telemetry.py", line 151, in wrap\n    result = func(*args, **kwargs)\n'
9. INSERT  INTO SNOWPARK_TEMP_TABLE_44LOL2QE6K("__index__", "A", "B", "C", "__row_position__") VALUES (?, ?, ?, ?, ?)
10. alter session unset query_tag
11. CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_44LOL2QE6K("__index__" BIGINT, "A" DOUBLE, "B" DOUBLE, "C" DOUBLE, "__row_position__" BIGINT)
12. alter session set query_tag = '  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/frame.py", line 706, in num_rows\n    return count_rows(self.ordered_dataframe)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/utils.py", line 1668, in count_rows\n    df.select(df.row_count_snowflake_quoted_identifier)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/_internal/telemetry.py", line 151, in wrap\n    result = func(*args, **kwargs)\n'
13. INSERT  INTO SNOWPARK_TEMP_TABLE_44LOL2QE6K("__index__", "A", "B", "C", "__row_position__") VALUES (?, ?, ?, ?, ?)
14. alter session unset query_tag
15. SELECT count(1) OVER (  ) AS "__row_count__" FROM ( SELECT...
16. DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_44LOL2QE6K
17. DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_44LOL2QE6K
```

with dedup
```
0. ls '@"TESTDB_SNOWPARK_PYTHON"."TESTSCHEMA_SNOWPARK_PYTHON".SNOWPARK_TEMP_STAGE_Y9M0URDKK1'
1. SELECT "name" FROM ( SELECT  *  FROM  TABLE ( RESULT_SCAN('01b67b31-0001-9f56-0000-047d100f939a')))
2. CREATE TEMPORARY  FUNCTION  "TESTDB_SNOWPARK_PYTHON"."TESTSCHEMA_SNOWPARK_PYTHON".SNOWPARK_TEMP_TABLE_FUNCTION_1AHR3RQ86H...
3. CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_Y9C70KYN3S("__index__" BIGINT, "A" DOUBLE, "B" DOUBLE, "C" DOUBLE, "__row_position__" BIGINT)
4. INSERT  INTO SNOWPARK_TEMP_TABLE_Y9C70KYN3S("__index__", "A", "B", "C", "__row_position__") VALUES (?, ?, ?, ?, ?)
5. CREATE  TEMPORARY  TABLE  "TESTDB_SNOWPARK_PYTHON"."TESTSCHEMA_SNOWPARK_PYTHON"."SNOWPARK_TEMP_TABLE_2S5501TCU3"("__row_position__" BIGINT, "LABEL" STRING(16777216), "VALUE" VARIANT)   AS  SELECT ...
6. DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_Y9C70KYN3S
7. CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_Y9C70KYN3S("__index__" BIGINT, "A" DOUBLE, "B" DOUBLE, "C" DOUBLE, "__row_position__" BIGINT)
8. alter session set query_tag = '  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/frame.py", line 706, in num_rows\n    return count_rows(self.ordered_dataframe)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/utils.py", line 1668, in count_rows\n    df.select(df.row_count_snowflake_quoted_identifier)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/_internal/telemetry.py", line 151, in wrap\n    result = func(*args, **kwargs)\n'
9. INSERT  INTO SNOWPARK_TEMP_TABLE_Y9C70KYN3S("__index__", "A", "B", "C", "__row_position__") VALUES (?, ?, ?, ?, ?)
10. alter session unset query_tag
11. SELECT count(1) OVER (  ) AS "__row_count__" FROM ( SELECT ...
12. DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_Y9C70KYN3S
```

File: tests/integ/modin/frame/test_iloc.py [hunk 4]
```diff
@@ -1052,13 +1052,13 @@ def iloc_helper(df):
         )
 
     high_count_reason = """
-        11 queries includes 5 queries to prepare the temp table for df, including create, insert,
-        drop the temp table (3.). and alter session to set and unset query_tag (2) and one select query.
+        6 queries includes queries to create, insert, and drop the temp table (3), alter session
+        to set and unset query_tag (2) and one select query.
```
Comment (sfc-gh-aalam) [inline tests/integ/modin/frame/test_iloc.py:1056]:
```
0: CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_5GYC8B2HIQ("__c0__" BIGINT, "c1" BIGINT, "__row_position__" >
1: alter session set query_tag = '  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/ut>
2: INSERT  INTO SNOWPARK_TEMP_TABLE_5GYC8B2HIQ("__c0__", "c1", "__row_position__") VALUES (?, ?, ?)
3: alter session unset query_tag
4: SELECT "__c0__", "c1" FROM ( SELECT  *  FROM (( SELECT "__index__" AS "__index__", "int_value_jpcc" AS "int_value_jpcc", "__r>
5: DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_5GYC8B2HIQ

---------------------------------------------------------------------

0: CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_NU1M8EN9NU("__index__" BIGINT, "__reduced__" DOUBLE, "__row_position__" BIGINT)
1: alter session set query_tag = '  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/utils.py", line 1332, in snowpark_to_pandas_helper\n    pandas_df = ordered_dataframe.to_pandas(statement_params=statement_params, **kwargs)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/ordered_dataframe.py", line 1740, in to_pandas\n    return snowpark_dataframe.to_pandas(\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/_internal/telemetry.py", line 151, in wrap\n    result = func(*args, **kwargs)\n'
2: INSERT  INTO SNOWPARK_TEMP_TABLE_NU1M8EN9NU("__index__", "__reduced__", "__row_position__") VALUES (?, ?, ?)
3: alter session unset query_tag
4: CREATE  OR  REPLACE  SCOPED TEMPORARY  TABLE SNOWPARK_TEMP_TABLE_C3E4S98KC9("__c0__" BIGINT, "c1" BIGINT, "__row_position__" BIGINT)
5: alter session set query_tag = '  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/utils.py", line 1332, in snowpark_to_pandas_helper\n    pandas_df = ordered_dataframe.to_pandas(statement_params=statement_params, **kwargs)\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/modin/plugin/_internal/ordered_dataframe.py", line 1740, in to_pandas\n    return snowpark_dataframe.to_pandas(\n  File "/Users/aalam/Projects/snowpark-python/src/snowflake/snowpark/_internal/telemetry.py", line 151, in wrap\n    result = func(*args, **kwargs)\n'
6: INSERT  INTO SNOWPARK_TEMP_TABLE_C3E4S98KC9("__c0__", "c1", "__row_position__") VALUES (?, ?, ?)
7: alter session unset query_tag
8: SELECT "__c0__", "c1" FROM ( SELECT  *  FROM (( SELECT "__index__" AS "__index__", "int_value_uwjn" AS "int_value_uwjn", "__row_position__" AS "__row_position__", "__row_position___2nnd" AS "__row_position___2nnd" FROM ( SELECT "__index__", iff(("int_value" < 0 :: INT), ("int_value" + "__row_count__"), "int_value") AS "int_value_uwjn", "__row_position__", "__row_position___2nnd" FROM ( SELECT  *  FROM (( SELECT "__index__" AS "__index__", "int_value" AS "int_value", "__row_position__" AS "__row_position__" FROM ( SELECT "__index__", trunc("__reduced___1v84", 0) AS "int_value", "__row_position__" FROM ( SELECT "__index__", "__reduced__", "__row_position__", "__reduced__" AS "__reduced___1v84" FROM ( SELECT  *  FROM (SNOWPARK_TEMP_TABLE_NU1M8EN9NU))))) AS SNOWPARK_LEFT CROSS JOIN ( SELECT "__row_position_ouyp__" AS "__row_position_ouyp__", "__row_count__" AS "__row_count__", "__row_position___2nnd" AS "__row_position___2nnd" FROM ( SELECT (row_number() OVER (  ORDER BY "__row_position__" ASC NULLS LAST ) - 1 :: INT) AS "__row_position_ouyp__", "__row_count__", "__row_position__" AS "__row_position___2nnd" FROM ( SELECT "__c0__", "c1", "__row_position__", count(1) OVER (  ) AS "__row_count__" FROM ( SELECT  *  FROM (SNOWPARK_TEMP_TABLE_C3E4S98KC9)) LIMIT 1))) AS SNOWPARK_RIGHT)))) AS SNOWPARK_LEFT INNER JOIN ( SELECT "__c0__" AS "__c0__", "c1" AS "c1", "__row_position___6i0y" AS "__row_position___6i0y" FROM ( SELECT "__c0__", "c1", "__row_position__" AS "__row_position___6i0y" FROM ( SELECT  *  FROM (SNOWPARK_TEMP_TABLE_C3E4S98KC9)))) AS SNOWPARK_RIGHT ON EQUAL_NULL("int_value_uwjn", "__row_position___6i0y"))) ORDER BY "__row_position__" ASC NULLS LAST, "__row_position___2nnd" ASC NULLS LAST, "__row_position___6i0y" ASC NULLS LAST
9: DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_NU1M8EN9NU
0: DROP  TABLE  If  EXISTS SNOWPARK_TEMP_TABLE_C3E4S98KC9
```

**Review by sfc-gh-yzou — APPROVED** [review]

**Pull request merged**